Release restrictions for BPFHostrouting when using the IPAMDelegatedPlugin

weiwei02 opened the pull request:
The IPAMDelegatedPlugin plugin makes it possible not to write 
the code of the cloud vendor into the core code of Cilium. 
This mode is conducive to the scalability of Cilium. 
However:

1. DaemonConfig check requires that 
`LocalRouterIPv4`/`LocalRouterIPv6` must be set when 
IPAMDelegatedPlugin is used.
2. When `LocalRouterIPv6`/`LocalRouterIPv6` is set, 
the verifier requires that the `EnableEndpointRoutes=true` 
option must be set.
3. which is equivalent to prohibiting the use of BPFHostrouting by DelegateIPAM.

After turning off the `EnableEndpointRoutes` option, cloud providers 
actually have the opportunity to set their own routes. 

For example, in the cni plugin chain, the private cni plugin of the 
cloud provider manages the local routing, and cooperates with `cilium-cni `
to implement BPFHostRouting to accelerate the container network.

weiwei02 committed:
Release restrictions for BPFHostrouting when using the IPAMDelegatedPlugin

The IPAMDelegatedPlugin plugin makes it possible not to write
the code of the cloud vendor into the core code of Cilium.
This mode is conducive to the scalability of Cilium.
However:

1. DaemonConfig check requires that
`LocalRouterIPv4`/`LocalRouterIPv6` must be set when
IPAMDelegatedPlugin is used.
2. When `LocalRouterIPv6`/`LocalRouterIPv6` is set,
the verifier requires that the `EnableEndpointRoutes=true`
option must be set.
3. which is equivalent to prohibiting the use of BPFHostrouting by DelegateIPAM.

After turning off the `EnableEndpointRoutes` option, cloud providers
actually have the opportunity to set their own routes.

For example, in the cni plugin chain, the private cni plugin of the
cloud provider manages the local routing, and cooperates with `cilium-cni `
to implement BPFHostRouting to accelerate the container network.

Signed-off-by: weiwei.wang <weiwei02@vip.qq.com>

pippolo84 commented on daemon/cmd/daemon_main.go:
```diff
@@ -1466,8 +1466,15 @@ func initEnv() {
+				log.Warnf("use ipam mode %s without %s, cilium will not manage local endpoint routes.",
+					ipamOption.IPAMDelegatedPlugin, option.EnableEndpointRoutes)
+			} else {
+				log.Fatalf("Cannot specify %s or %s  without %s.", option.LocalRouterIPv4, option.LocalRouterIPv6, option.EnableEndpointRoutes)
```
Nit: additional space
```suggestion
				log.Fatalf("Cannot specify %s or %s without %s.", option.LocalRouterIPv4, option.LocalRouterIPv6, option.EnableEndpointRoutes)
```

pippolo84 reviewed (approved):
LGTM apart a very small nit in the logging message.
Looped in @gandro to double check IPAM related changes.

weiwei02 committed:
Update daemon/cmd/daemon_main.go

Co-authored-by: Fabio Falzoi <fabio.falzoi84@gmail.com>
Signed-off-by: weiwei.wang <weiwei02@vip.qq.com>

gandro reviewed (changes requested):
Thanks for the PR!

I'm generally not a fan of using a warning log message here. A warning should always be designed in a way where the user can fix something to make the warning go away.

Therefore, I think it would be better to add a flag instead, e.g. `--externally-managed-endpoint-routes` (there is probably a better name) for the user to indicate that the routes are managed by some external system.

@Weil0ng Could you please take a look as well, since you were the one who added the requirement to require EndpointRoutes.

weiwei02 committed:
agent: add externally managed endpoint routes

Add the 'externally-managed-endpoint-routes' option for the cimium
agent, which can only be used with IPAMDelegatedPlugin.
When 'externally-managed-endpoint-routes' is enabled, it indicates
that in IPAMDelegatedPlugin, endpoints routes are all managed by
external systems which have routing management capability.

Signed-off-by: weiwei.wang <weiwei02@vip.qq.com>

gandro reviewed (commented):
Thanks for the update, this looks good to me - assuming we actually need the old EndpointRoutes check (waiting on @Weil0ng to chime in if possible).
